File: src/template/cpp/dsu/dsu.cpp
#include <vector>
using namespace std;

struct DSU {
    vector<int> p, sz;
    DSU(int n): p(n), sz(n,1) {
        for(int i = 0; i < n; ++i) p[i] = i;
    }
    int find(int x) {
        return p[x] == x ? x : p[x] = find(p[x]);
    }
    bool unite(int a, int b) {
        a = find(a); b = find(b);
        if (a == b) return false;
        if (sz[a] < sz[b]) swap(a, b);
        p[b] = a;
        sz[a] += sz[b];
        return true;
    }
    bool same(int a, int b) {
        return find(a) == find(b);
    }
};

File: src/template/julia/dsu/dsu.jl
struct DSU
    parent::Vector{Int}
    size::Vector{Int}
end

function DSU(n::Int)
    DSU(collect(1:n), ones(Int, n))
end

function find(d::DSU, x::Int)
    p = d.parent
    while p[x] != x
        p[x] = p[p[x]]
        x = p[x]
    end
    return x
end

function unite!(d::DSU, a::Int, b::Int)
    a = find(d, a); b = find(d, b)
    if a == b
        return false
    end
    if d.size[a] < d.size[b]
        a, b = b, a
    end
    d.parent[b] = a
    d.size[a] += d.size[b]
    return true
end

same(d::DSU, a::Int, b::Int) = find(d, a) == find(d, b)